SQL/proc_descriptions/testing_extension.sql

```diff
@@ -0,0 +1,253 @@
+    result_record RECORD;
+BEGIN
+    -- A little note about raising exception in this function, it seems that `_runner` function
+    -- does not raise an exception, instead catching it and returning it as a record, so under
+    -- this assumption we raise "division by zero" AKA "22012" exception to rollback the transaction.
+    BEGIN -- begin transaction for later rollback
+        FOR result_record IN
+            SELECT * FROM _runner(
+                findfuncs_recursive( $1, startup_pattern ),
+                findfuncs_recursive( $1, shutdown_pattern ),
+                findfuncs_recursive( $1, setup_pattern ),
+                findfuncs_recursive( $1, teardown_pattern ),
+                findfuncs_recursive( $1, test_pattern, exclude_pattern )
+            )
+        LOOP
+            RETURN NEXT result_record;
+        END LOOP;
+
+        -- Raise the exception
+        RAISE EXCEPTION '__TAP_ROLLBACK__' USING ERRCODE = '22012';
+    EXCEPTION
+        WHEN SQLSTATE '22012' THEN
+            -- Catch the exception to prevent it from propagating
+            NULL; -- Do nothing
+    END;
+END;
+$$;
+
+-- Function: mob_group_runtests(text)
+-- This function is used to run a group of tests.
+-- Input parameters: text - the pattern to match the names of the functions to run.
+-- Flow:
+-- 1. Run alphabetically ordered startup functions once, which match the pattern "startup" + $1.
+-- 2. Enter a loop:
+--  2.1. Run alphabetically ordered setup functions once, which match the pattern "setup" + $1.
+--  2.2. Run alphabetically ordered test functions once, which match the pattern $1.
+--  2.3. Run alphabetically ordered teardown functions once, which match the pattern "teardown" + $1.
+-- 3. Run alphabetically ordered shutdown functions once, which match the pattern "shutdown" + $1.
+-- 4. Return the results of the tests.
+create or replace function mob_group_runtests(text) returns SETOF text
+    language plpgsql
+as
+$$
+DECLARE
+    record RECORD;
+BEGIN
+    FOR record IN
+        SELECT * FROM mob_group_runtests(NULL, $1)
+    LOOP
+        RETURN NEXT record;
+    END LOOP;
+END;
+$$;
+
+-- procedures for creating testing environment
+
+CREATE OR REPLACE PROCEDURE test_env_constructor(text) AS
+$$
+DECLARE
+    test_scheme_name TEXT := $1;
+    table_name TEXT;
+BEGIN
+    -- check if schema exists
+    IF EXISTS (SELECT 1 FROM information_schema.schemata WHERE schema_name = test_scheme_name) THEN
+        RAISE EXCEPTION 'Schema % already exists', test_scheme_name;
+    END IF;
+
+    -- Create schema test_scheme_name
+    CREATE SCHEMA test_scheme_name;
+
+    -- copy tables with no data to test_scheme_name
+    FOR table_name IN (SELECT table_name FROM information_schema.tables WHERE table_schema = 'public')
+        LOOP
+            EXECUTE format('CREATE TABLE %I.%I AS TABLE public.%I WITH NO DATA', test_scheme_name, table_name, table_name);
```

Comment:
@F-I-D-O look at the calls of `CREATE SCHEMA ...` and `CREATE TABLE ...`. I run it by non-superuser in my local db and it showed warning for insufficient permissions. That means, that if we will integrate this functions into `mob_group_runtests()` common users will not be able to run tests. Im not sure how to proceed

Reply:
I think we can use some fixed test schema for tests. Create table should be enabled in that schema for all users.